src/psyclone/psyir/backend/sympy_writer.py

```diff
@@ -115,86 +132,212 @@ def __init__(self, type_map=None):
+            specified, or a list of SymPy expressions.
+        :rtype: Union[:py:class:`psyclone.psyir.backend.SymPyWriter`,
+                      List[:py:class:`sympy.core.basic.Basic`]
+
         '''
-        This function creates a dictionary mapping each Reference in any
-        of the expressions to either a Sympy Function (if the reference
+        if expressions:
+            # If we have parameters, create an instance of the writer
+            # and use it to convert the expressions:
+            writer = SymPyWriter()
+            return writer(expressions)
+
+        # No parameter, just create an instance and return it:
+        return super().__new__(cls)
+
+    # -------------------------------------------------------------------------
+    def __getitem__(self, _):
+        '''This function is only here to trick pylint into thinking that
+        the object returned from ``__new__`` is subscriptable, meaning that
+        code like:
+        ``out = SymPyWriter(exp1, exp2); out[1]`` does not trigger
+        a pylint warning about unsubscriptable-object.
+        '''
+
+    # -------------------------------------------------------------------------
+    def _create_type_map(self, list_of_expressions):
+        '''This function creates a dictionary mapping each Reference in any
+        of the expressions to either a SymPy Function (if the reference
         is an array reference) or a Symbol (if the reference is not an
-        array reference).
+        array reference). It defines a new SymPy function for each array,
+        which has a special write method implemented that automatically
+        converts array indices back by combining each three arguments into
+        one expression (i. e. ``a(1,9,2)`` would become ``a(1:9:2)``).
+
+        A new symbol table is created any time this function is called, so
+        it is important to provide all expressions at once for the symbol
+        table to avoid name clashes in any expression.
 
-        :param list_of_expressions: the list of expressions from which all \
-            references are taken and added to the a symbol table to avoid \
+        :param list_of_expressions: the list of expressions from which all
+            references are taken and added to the a symbol table to avoid
             renaming any symbols (so that only member names will be renamed).
-        :type list_of_expressions: list of \
-            :py:class:`psyclone.psyir.nodes.Node`
-        :returns: the dictionary mapping each reference name to a Sympy \
-            data type (Function of Symbol).
-        :rtype: dictionary of string:Sympy-data-type values
+        :type list_of_expressions: List[:py:class:`psyclone.psyir.nodes.Node`]
 
         '''
-        sympy_type_map = {}
+        # Create a new symbol table, so previous symbol will not affect this
+        # new conversion (i.e. this avoids name clashes with a previous
```

Comment:
I gathered from the doc string that name clashes with previous conversions are possible because we create a new table each time? Perhaps we should pass in a table?

Reply:
OK.